test/tstDetailsTreeNodeLabelling.cpp

```diff
@@ -0,0 +1,204 @@
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "ArborX_EnableDeviceTypes.hpp" // ARBORX_DEVICE_TYPES
+#include "ArborX_EnableViewComparison.hpp"
+#include <ArborX_DetailsTreeNodeLabelling.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+#include "BoostTest_CUDA_clang_workarounds.hpp"
+#include <boost/test/unit_test.hpp>
+
+#include <random>
+
+template <typename DeviceType, typename T>
+auto toView(std::vector<T> const &v, std::string const &lbl = "")
+{
+  Kokkos::View<T *, DeviceType> view(lbl, v.size());
+  Kokkos::deep_copy(view, Kokkos::View<T const *, Kokkos::HostSpace,
+                                       Kokkos::MemoryTraits<Kokkos::Unmanaged>>(
+                              v.data(), v.size()));
+  return view;
+}
+
+template <class MemorySpace>
+struct MockBVH
+{
+  Kokkos::View<Kokkos::pair<int, int> *, MemorySpace> children_;
+  KOKKOS_FUNCTION auto size() const { return children_.extent(0) + 1; }
+};
+
+#define HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(MEMORY_SPACE)           \
+  template <>                                                                  \
+  auto ArborX::Details::HappyTreeFriends::getLeftChild<MockBVH<MEMORY_SPACE>>( \
+      MockBVH<MEMORY_SPACE> const &x, int i)                                   \
+  {                                                                            \
+    return x.children_(i).first;                                               \
+  }                                                                            \
+  template <>                                                                  \
+  auto                                                                         \
+  ArborX::Details::HappyTreeFriends::getRightChild<MockBVH<MEMORY_SPACE>>(     \
+      MockBVH<MEMORY_SPACE> const &x, int i)                                   \
+  {                                                                            \
+    return x.children_(i).second;                                              \
+  }
+
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(Kokkos::HostSpace)
+#ifdef KOKKOS_ENABLE_CUDA
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(Kokkos::Cuda::memory_space)
+#endif
+#ifdef KOKKOS_ENABLE_HIP
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(
+    Kokkos::Experimental::HIP::memory_space)
+#endif
+#ifdef KOKKOS_ENABLE_SYCL
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(
+    Kokkos::Experimental::SYCL::memory_space)
+#endif
+#ifdef KOKKOS_ENABLE_OPENMPTARGET
+HAPPY_TREE_FRIENDS_GET_CHILDREN_SPECIALIZATION(
+    Kokkos::Experimental::OpenMPTarget::memory_space)
+#endif
```

Comment:
Hmm, this is painful. It also likely misses stuff (for example, if both `Serial` and `OpenMP` are available). We'll also have to remember to update it when adding a new space. Have you tried using a templated specialization of `getLeftChild`, or does not not play together with the original `HappyTreeFriends`, which is templated on `BVH`?

Reply:
> It also likely misses stuff (for example, if both Serial and OpenMP are available).

They both use `Kokkos::HostSpace`

Reply:
Note to self: functions do not allow partial specializations. Forgot about that.